File: src/main/java/com/rephilo/luandun/leetcode/_476/Solution.java
package com.rephilo.luandun.leetcode._476;

public class Solution {
    /**
     * 476
     * 位运算确实是弱项，需要再学习
     *
     * @param num
     * @return
     */
    public int findComplement(int num) {
        int temp = num, c = 0;
        while (temp > 0) {
            //根据判断条件
            //二进制右移并赋值给temp，
            temp >>= 1;
            //二进制左移之后结果+1 赋值给c
            c = (c << 1) + 1;
        }
        return num ^ c;
    }
}

File: src/main/scala/com/rephilo/luandun/leetcode/_476/Solution.scala
package com.rephilo.luandun.leetcode._476

import scala.annotation.tailrec

object Solution {
  /**
    * 476
    * 不知道为啥执行不下去呀
    *
    * @param num
    * @return
    */
  def findComplement(num: Int): Int = {
    @tailrec
    def calc(num: Int, c: Int): Int = {
      num match {
        case _ if (num <= 0) => num ^ c
        case _ => calc(num >> 1, (c << 1) + 1)
      }
    }

    calc(num, 0)
  }
}
